ara_plumes/tests/test_concentric_circle.py

```diff
@@ -0,0 +1,208 @@
+from ..concentric_circle import _interpolate_intersections
+from ..concentric_circle import concentric_circle
+from ..models import get_contour
+from ..typing import X_pos
+from ..typing import Y_pos
+
+
+def test_concentric_circle():
+    # create data
+    scale = 4
+    orig_center = (X_pos(50), Y_pos(200))
+    height, width = 400, 400
+    bw_img = np.zeros((height, width), dtype=np.uint8)
+
+    bw_img[25:376, 50] = 255 // scale
+    bw_img[25, 50:251] = 255 / scale
+    bw_img[375, 50:251] = 255 // scale
+    bw_img[25:376, 250] = 255 // scale
+
+    num_of_circs = 3
+    r = 50
+    for i in range(num_of_circs):
+        i += 2
+        bw_img[200, r * i] = 255
+
+    rectangle_plume = bw_img.copy()
+
+    # get result
+    selected_contours = get_contour(rectangle_plume)
+    result_mean, result_var1, result_var2 = concentric_circle(
+        rectangle_plume,
+        contours=selected_contours,
+        orig_center=orig_center,
+        radii=r,
+        num_of_circs=num_of_circs,
+    )
+
+    # expected data
+    expected_mean = np.array(
+        [[0, 50, 200], [50, 100, 200], [100, 150, 200], [150, 200, 200]]
+    )
+    expected_var1 = np.array(
+        [[0, 50, 200], [50, 50, 250], [100, 50, 300], [150, 50, 350]]
+    )
+    expected_var2 = np.array(
+        [[0, 50, 200], [50, 50, 150], [100, 50, 100], [150, 50, 50]]
+    )
+
+    np.testing.assert_array_equal(expected_mean, result_mean)
+    np.testing.assert_array_equal(expected_var1, result_var1)
+    np.testing.assert_array_equal(expected_var2, result_var2)
+
+
+def test_find_max_on_boundary():
+    scale = 4
+    orig_center = (50, 200)
+    height, width = 400, 400
+    bw_img = np.zeros((height, width), dtype=np.uint8)
+
+    bw_img[25:376, 50] = 255 // scale
+    bw_img[25, 50:251] = 255 / scale
+    bw_img[375, 50:251] = 255 // scale
+    bw_img[25:376, 250] = 255 // scale
+
+    num_of_circs = 3
+    r = 50
+    for i in range(num_of_circs):
+        i += 2
+        bw_img[200, r * i] = 255
+
+    rectangle_plume = bw_img.copy()
+
+    expected = (255, np.array([100, 200]))
+    result = _find_max_on_boundary(rectangle_plume, orig_center, r=50)
+
+    np.testing.assert_equal(expected[0], result[0])
+    np.testing.assert_array_almost_equal(expected[1], result[1])
+
+
+def test_get_edge_points_3_circles():
+    num_of_circs = 3
+    radii = 1
+    contour_distances = [
+        np.array(
+            [
+                [np.sqrt(2), 0, 0],
+                [np.sqrt(2), 2, 0],
+                [np.sqrt(10), 2, 2],
+                [np.sqrt(10), 0, 2],
+            ]
+        )
+    ]
+    orig_center = (1, -1)
+    result = _get_edge_points(num_of_circs, radii, contour_distances, orig_center)
+    expected = (
+        np.array([[2, 0, np.sqrt(3) - 1], [3, 0, 2 * np.sqrt(2) - 1]]),
+        np.array([[2, 2, np.sqrt(3) - 1], [3, 2, 2 * np.sqrt(2) - 1]]),
+    )
+    for e, r in zip(expected, result):
+        np.testing.assert_array_almost_equal(e, r)
```

Comment:
Since this is a bit higher-level of an integration test, rather than a unit test, it's trickier to reason about the exact expected result.  Instead of `assert_array_almost_equal`, can we just test the stuff  `_get_edge_points` is responsible for, assuming all the lower-level functions work perfectly?
- [ ] that `len(contour_crosses)==0` causes a skip (e.g. `assert len(ccw) == 2`)
- [ ] that the cw_edge_points are all clockwise of the counterclockwise edge points.

Reply:
I like this a lot! Have modified test. Also included link (as comment in test) to a visual of the test that was created.